src/algebra/char_p/mixed_char_zero.lean

```diff
@@ -0,0 +1,471 @@
+after we reduced to `p` prime, we can assume that the ideal `I` in the definition is maximal.
+-/
+lemma reduce_to_maximal_ideal {p : ℕ} (hp : nat.prime p) :
+  (∃ (I : ideal R), (I ≠ ⊤) ∧ char_p (R ⧸ I) p) ↔
+  (∃ (I : ideal R), (I.is_maximal) ∧ char_p (R ⧸ I) p) :=
+begin
+  split,
+  { intro g,
+    rcases g with ⟨I, ⟨hI_not_top, hI⟩⟩,
+
+    -- Krull's Thm: There exists a prime ideal `M` such that `I ≤ M`.
+    rcases ideal.exists_le_maximal I hI_not_top with ⟨M, ⟨hM_max, hM⟩⟩,
+
+    use M,
+    split,
+    exact hM_max,
+    { cases char_p.exists (R ⧸ M) with r hr,
+      convert hr,
+      resetI, -- make `hr : char_p (R ⧸ M) r` an instance.
+
+      have r_dvd_p : r ∣ p :=
+      begin
+        rw ←char_p.cast_eq_zero_iff (R ⧸ M) r p,
+        convert congr_arg (ideal.quotient.factor I M hM) (char_p.cast_eq_zero (R ⧸ I) p)
+      end,
+      rw eq_comm,
+      apply or_iff_not_imp_left.mp (nat.prime.eq_one_or_self_of_dvd hp r r_dvd_p),
+      exact char_p.char_ne_one (R ⧸ M) r }},
+  { intro g,
+    rcases g with ⟨I, ⟨hI_max, hI⟩⟩,
+    use I,
+    exact ⟨ideal.is_maximal.ne_top hI_max, hI⟩ }
+end
+
+end mixed_char_zero
+
+/-!
+### Equal characteristic zero
+
+A commutative ring `R` has "equal characteristic zero" if it satisfies one of the following
+equivalent properties:
+
+1) `R` is a `ℚ`-algebra.
+2) The quotient `R ⧸ I` has characteristic zero for any proper ideal `I ⊂ R`.
+3) `R` has characteristic zero and does not have mixed characteristic for any prime `p`.
+
+We show `(1) ↔ (2) ↔ (3)`, and most of the following is concerned with constructing
+an explicit algebra map `ℚ →+* R` (given by `x ↦ (x.num : R) /ₚ ↑x.pnat_denom`)
+for the direction `(1) ← (2)`.
+
+Note: Property `(2)` is denoted as `equal_char_zero` in the statement names below due to lack
+of a nice description.
+-/
+section equal_char_zero
+
+/--
+`ℚ`-algebra implies equal characteristic.
+-/
+lemma Q_algebra_to_equal_char_zero [nontrivial R] [algebra ℚ R] :
+  ∀ (I : ideal R), I ≠ ⊤ → char_zero (R ⧸ I) :=
+begin
+    haveI : char_zero R := @char_p.char_p_to_char_zero R _
+      (char_p_of_injective_algebra_map (algebra_map ℚ R).injective 0),
+    intros I hI,
+    exact
+    ⟨begin
+      intros a b h_ab,
+      apply @nat.cast_injective R,
+      simp_rw ←map_nat_cast (ideal.quotient.mk I) at h_ab,
+      simp_rw ←map_nat_cast (algebra_map ℚ R) at ⊢ h_ab,
+      apply congr_arg,
+      rw ←sub_eq_zero at ⊢ h_ab,
+      simp_rw ←map_sub at ⊢ h_ab,
+      rw ideal.quotient.eq_zero_iff_mem at h_ab,
+
+      -- `i(↑a - ↑b)` is a unit contained in `I`, which contradicts `I ≠ ⊤`.
+      by_contradiction h_ne_zero,
+      have hI' : I = ⊤ :=
+        I.eq_top_of_is_unit_mem h_ab (is_unit.map (algebra_map ℚ R) (is_unit.mk0 _ h_ne_zero)),
+      exact absurd hI' hI,
+    end⟩
+end
+
+section construction_of_Q_algebra
+
+/-- Internal: Not intended to be used outside this local construction. -/
```

Comment:
Is there a better way to mark lemmas that are only used locally? With the [fact ...] they are only used to get the construction in equal_char_zero_to_Q_algebra.

Reply:
In practice, what you did is the best way. In theory, you could make it `protected`, but that is discouraged because it has happened too often that someone outside this file still finds a use for your "internal" lemma.